prover/scripts/test_prover

```diff
@@ -1,36 +1,61 @@
 #!/usr/bin/env bash
 
-# Basic tests of the prover commands, primarily to check consistency against
-# the SDK and contract code.
+# Tests keygen, proving and verification
+
+set -e
+set -x
 
 # Build
-pushd ../upa/
-  yarn
-  yarn build
-  yarn
-popd
 cargo build --release
 
 # Shell setup
 shopt -s expand_aliases
 . ./scripts/shell_setup.sh
 
-set -e
-set -x
+DRY_RUN_FLAG=""
+if [ "$DRY_RUN" -eq 1 ]; then
+  DRY_RUN_FLAG="--dry-run"
+fi
 
+rm -rf _test_prover
 mkdir -p _test_prover
 pushd _test_prover
 
-    cp ../../circuits/src/tests/data/vk.json vk.json
+    # Get a default config if one is not present.
+    [ -e upa_config.json ] || cp ../tests/upa_config_2.json upa_config.json
+    inner_batch_size=`jq -r .inner_batch_size upa_config.json`
+    max_num_inputs=`jq -r .max_num_app_public_inputs upa_config.json`
+    outer_batch_size=`jq -r .outer_batch_size upa_config.json`
+
+    # Create `inner_batch_size` (unsafe) keys.
+    for i in $(seq 0 $(($inner_batch_size - 1))); do
+      key_file="vk-${i}.unsafe.json"
+      if [ $(($i % 2)) -eq 0 ]; then
+        prover groth16 generate-fake-vk -i ${max_num_inputs} -v $key_file
+      else
+        prover groth16 generate-fake-vk -i $((${max_num_inputs} / 2)) -v $key_file -c
+      fi
+    done
+
+    # Create `outer_batch_size` inner batches
+    for j in $(seq 0 $(($outer_batch_size - 1))); do
+      prover groth16 generate-proofs \
+        -n ${inner_batch_size} \
+        $(for i in $(seq 0 $(($inner_batch_size - 1))); do echo -v vk-${i}.unsafe.json; done) \
+        -b inner-batch-${j}.json
+    done
+
+    batch_files=""
+    for j in $(seq 0 $(($outer_batch_size - 1))); do
+      batch_files+="--app-vk-proof-batch inner-batch-${j}.json "
+    done
+    echo ${batch_files}
 
-    # Check native and Typescript cids match
-    cid_evm=`upa compute circuit-id vk.json`
-    cid_native=`prover universal-batch-verifier compute-circuit-id --app-vk=vk.json`
-    if ! [ "${cid_evm}" == "${cid_native}" ] ; then
-        echo "cid_evm: ${cid_native}"
-        echo "cid_native: ${cid_native}"
-        exit 1
-    fi
+    # Run keygen
+    dummy_srs_setup
+    prover full keygen ${DRY_RUN_FLAG}
```

Comment:
This seems to just replcate what `keygen` does.

Reply:
Removed the `keygen` script. This will imply some changes in saturn, which I'll make in the corresponding PR. I simply believe it's cleaner (and easier to maintain) if we keep as many commands as possible in the rust side.

Reply:
I think it is alright to add `full keygen` in Rust, in particular if its useful for people who want to play around with the open source tool.

But IMO, we should not fully remove `keygen` bash script because it generates the outer verifier contract with `solc` and does useful error checking around this process. We _could_ split this step into a separate bash script. But I think most of the team has gotten used to `keygen` script doing this for us.

Reply:
Yeah, I do know what you mean about `keygen`, although I don't like having two implementations (and two ways for users) to do the same thing.

OK, let's take the easy path for now then and keep both.  It will save having to replicate all the behavior of `keygen` in rust.  It feels like technical debt for now reason, but let's just get this in so we can move on, and sort it later.

Reply:
In that case, I'll restore it. 

Afaik, the command `solc --yul ${YUL_CODE} --bin` produces the same as the `snark_verifier` counterpart `compile_yul`. So it should be easy to get rid of the `keygen` script later.